### PR TITLE
Introducing TrickOps - An Extensible Sim Testing Framework

### PR DESCRIPTION
**Pull request by ddj116**
Features:

* Multiple simultaneous sim builds, runs, file vs. file comparisons,
  arbitrary post-run analyses, valgrind of runs
* Real-time progress bars for sim builds and runs
* Exit code management lets users easily define success & failure
* Failed comparisons can optionally, generate koviz error reports

TODO:

* Finish unit and doc tests
* Flesh out README.md documentation
* Code review from Trick team

### CODE REVIEW

**Comment by ddj116** [comment]
@spfennell @jmpenn @alexlin0 @keithvetter  -- here's the code associated with the presentation I sent you. I've got about 20 unit tests and a handful of doctests so far, but will make more soon. I've got a Gateway project referencing this PR source branch that I'm collaborating on with another person now, so I wanted to get something published even though this is still WIP.  I believe the code is ready to be looked at though, just writing more tests/docs is left. A few key points that will be better fleshed out in the forthcoming `README.md`

* This code requires `python3` and the packages mentioned in `share/trick/trickops/requirements.txt`
* To run the unit tests: `cd share/trick/trickops/test && ./run_tests.py`
* To run an example workflow using a trick clone in `/tmp/`: `cd share/trick/trickops/ && ./ExampleWorkflow.py`

**Comment by ddj116** [comment]
Forgot to @dbankieris !

**Comment by ddj116** [comment]
OK @jmpenn I've got extensive documentation per your request, see the `README.md` file in this PR [here](https://github.com/nasa/trick/pull/1130/files#diff-fdb63e2024a318cbba124fc6b1e18fa5cc84dbde8a63451e0f0a010acffd2f82)

I believe this is ready for code review if y'all have the time. There are 35 python unit tests and 80 doctests all of which are passing. I'll work with @spfennell on how exactly to fold these into GitHub actions.

File: share/trick/trickops/ColorStr.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        return self.CODE
+
+    def termcode(self, num):
+        return '\033[%sm'%num
+
+    def colorstr(self, astr, color):
+        return self.termcode(self.CODE[color])+astr+self.termcode(self.CODE['ENDC'])
+
+    def showCodes( self ):
+        astr='yippy skippy'
```
Comment (jmpenn) [inline share/trick/trickops/ColorStr.py:58]: :smirk:

Comment (ddj116) [reply share/trick/trickops/ColorStr.py:58]: lol this is copy-pasta that dates back years, can easily remove. Part of me wondered if `ColorStr` belongs in `WorkflowCommon.py` but I don't have a strong preference on it either way.

**Comment by jmpenn** [comment]
Dan, this'll be a very nice contribution to Trick. I'd like to give a little more time for others to chime in, then I'm ready to merge it. Thanks!

File: share/trick/trickops/ColorStr.py (hunk 1 above)
Comment (jmpenn) [reply share/trick/trickops/ColorStr.py:58]: Don't remove it! I like it!

**Comment by spfennell** [comment]
Just a general comment, trick_ops examples and unit tests use trick sims that are available in trick. Those sims were not created for trickops, but now If we want to add/modify/remove any of those sims, trickops needs to be modified as well. Could trickops have it's own sim directory with its own sims to improve encapsulation? In regards to both the example workflow and the unit tests(they use SIM_alloc_test and the ball sim)

**Comment by ddj116** [comment]
> Could trickops have it's own sim directory with its own sims to improve encapsulation?

Definitely possible, but wouldn't this require me to copy a lot of code / SIM content that already exists in `trick/`?  I understand what your getting at with respect to encapsulation, but isn't duplicating content in Trick an issue unto itself? What do you think?

>  but now If we want to add/modify/remove any of those sims, trickops needs to be modified as well. 

Excellent point. As you've likely seen `ExampleWorkflow.py` is reading the YAML file which is hardcoded in the file. If for example y'all rename a SIM and that YAML file no longer has the correct sim `path:`, then it will print a non-fatal error but continue along.  Here's one such example where I purposefully break the path to `SIM_ball_L1`:

```
ERROR: SIM_ball_L1's 'path' not found. Continuing but skipping this entire entry from /tmp/config.yml.
```

Additionally, since some runs in Trick fail for various reasons, we can't check the exit code of `ExampleWorkflow.py` in GitHub actions to determine success/failure on TrickOps itself. Indeed, `ExampleWorkflow.py` showing what happens when a sim fails is itself a success.  I actually wasn't quite sure if even running `ExampleWorkflow.py` in GitHub Actions was a good idea, since "success" is a bit hard to define. Really we'd need to check the `status_summary()` output matches what we expect:

```
SUMMARY:
  47 out of 54 builds succeeded
  28 out of 37 runs succeeded
```
Since there's a suite of unit tests whose success criteria is black-and-white, I was going to start with just the unit tests and consider `ExampleWorkflow.py` more documentation than an actual CI test. Let me know what you think.

---

Also -- now that there's some level of approval I just wanted to say please don't merge this yet. I've found some typos and other fixes I'd like to push in before this is merged proper. I'd also like to get a new GitHub Actions workflow live, hoping to have that working this week.

FYI @jmpenn @alexlin0 @spfennell @dbankieris just a heads-up -- I've formally requested via NAMS the ability to be given access to the trick project in Github. This PR-from-my-fork works for now, but its wrecks havok on our GitLab CI process which can't find the Trick submodule hash because it's not on a branch in `nasa/trick.git`. I don't need (or even want) the ability to push to or merge a PR into `master`, I just want the ability to push a topic branch so our CI process can see the hashes for our Trick submodule. Hope this is OK / possible, let me know if you want to discuss.

**Comment by ddj116** [comment]
OK GitHub Actions is live with the new TrickOps workflow in this PR. Successful output [here](https://github.com/ddj116/trick/actions/runs/793993287).  I'm only testing ubuntu 20 and centos 8 right now, I'm holding off on Mac since this job actually downloads and builds koviz and I don't quite know how to do that or if koviz is even stable for Mac.

I believe this is ready to merge whenever y'all are satisfied. Note that MacOS is failing at this state but I believe that's only because this branch is based on a Feb version of Trick where I suspect that test wasn't stable. Once merged I'd expect all tests currently passing in `master` to pass post-merge. Normally I'd rebase this work onto `origin/master`, but I'd like to purposefully *not* do that right now. Reason being is I don't yet know if our gateway project is stable in the latest trick, but I know we're stable at 5d7b6e3bcc3dd16ecbf5ad98ddac494d67d12c58 which is why I based my branch off of that commit.  We will of course step to the latest Trick `master` soon, but I didn't want to be forced into doing that as part of this work, if that makes sense.

Also, I don't know who needs to resolve the **1 workflow awaiting approval**, but I'm still seeing it on this PR.

**Comment by ddj116** [comment]
Actually hold off one sec, I inadvertently changed some workflow files by restoring the wrong version. Stand by.

**Comment by ddj116** [comment]
OK, I believe it's ready now but double check me on the Mac OS failure being unrelated to this work.  Also, documentation is linked from the top level and can be previewed in [my fork pages area]( http://ddj116.github.io/trick/documentation/miscellaneous_trick_tools/TrickOps.html#learn-trickops-with-an-example-workflow-that-uses-this-trick-repository)

**Comment by ddj116** [comment]
FYI I talked to John Penn today and after discussing with him, it sounds like yall are targeting merge of this PR on Monday. That's great and gives me time to implement three more changes:
1. Do not allow `valgrind` runs if `sys.platform` is Darwin since valgrind doesn't exist on Mac
2. Change `ExampleWorkflow.py` to a smaller subset of only stable sims/runs.
3. Add `ExampleWorkflow.py` to GitHub actions

After talking to John I changed my mind about `ExampleWorkflow.py` and not including it in GitHub Actions. At first I liked the idea of letting the user know what a failure looks like, but that's probably not worth the "why is trick broke???" emails that will surely come when folks run it.  Plus, if we test only stable sims/runs then I can continuously test it in GitHub Actions.

This should be pretty quick to implement, should be done by COB Friday.

**Comment by ddj116** [comment]
Regarding this list:

 1. Do not allow `valgrind` runs if `sys.platform` is Darwin since valgrind doesn't exist on Mac
 2. Change `ExampleWorkflow.py` to a smaller subset of only stable sims/runs.
 3. Add `ExampleWorkflow.py` to GitHub actions

(1) is complete. (2) is complete. (3) is deferred because I remembered that `ExampleWorfklow.py` won't run without a pre-built Trick, which these unit test jobs do not have. I'd like to figure out a good way to test this file, but that's a bigger effort and bringing in an entire Trick build breaks my "these are pure unit tests" assumption. I put a big fat `TODO` in the file explaining my thinking on this:

```yaml
# TODO: ExampleWorkflow.py is not included here because it needs a built Trick
#  to function correctly and I don't want to duplicate the Trick build testing
#  here to provide testing of what is essentially an example provided for
#  documentation purposes.  If we could leverage artifacts from a previous
#  stage and/or stable containers where Trick is already pre-built, we should
#  consider adding ExampleWorfklow.py to testing in this file. -Jordan 4/2021
```

The TrickOps tests are passing again, and this is ready to merge again from my perspective. Y'all have a great weekend!

**Comment by ddj116** [comment]
ACK please hold off on the merge, I found a logic error that I've got a fix for and will push soon.

**Comment by ddj116** [comment]
OK, pushed a fix to the logic error and an accompanying test that covers it. Good to go again.

**Pull request merged**